sycl/include/sycl/ext/oneapi/experimental/reduction_properties.hpp

```diff
@@ -83,9 +97,17 @@ template <typename BinaryOperation>
 struct IsDeterministicOperator<DeterministicOperatorWrapper<BinaryOperation>>
     : std::true_type {};
 
+template <typename PropertyList>
+inline constexpr bool is_valid_reduction_prop_list =
+    ext::oneapi::experimental::detail::all_are_properties_of_v<
+        ext::oneapi::experimental::detail::reduction_property_check_anchor,
+        PropertyList>;
+
 } // namespace detail
 
-template <typename BufferT, typename BinaryOperation, typename PropertyList>
+template <typename BufferT, typename BinaryOperation, typename PropertyList,
+          typename = std::enable_if_t<
+              detail::is_valid_reduction_prop_list<PropertyList>>>
 auto reduction(BufferT vars, handler &cgh, BinaryOperation combiner,
                PropertyList properties) {
```

Comment:
In this case, it seems like you could avoid adding a new template parameter by instead adding `enable_if` to the `handler` parameter.

In general, implementing constraints seems a bit hacky in C++17 because you need to use SFINAE, and this sometimes requires adding a new template parameter or even changing a non-templated member function into a templated one.  I don't know if we have a strict set of rules about whether this is completely legal.  We know that changing a conversion operator from non-templated to templated changes the semantics, so we don't do this.  (@aelovikov-intel has used a "mix-in" technique to avoid this for the `vec` class.)  For the other cases, I'd advise that we avoid adding new template parameters if there is an alternative.

Reply:
> by instead adding enable_if to the handler parameter.

I don't remember seeing such an approach anywhere in the codebase, do you have some pointers?

> For the other cases, I'd advise that we avoid adding new template parameters if there is an alternative.

I thought we're adding that extra parameter almost in every corner of the project right now...

Reply:
Can't you do something like this:

```
template <typename BufferT, typename BinaryOperation, typename PropertyList>
auto reduction(BufferT vars, std::enable_if<detail::is_valid_reduction_prop_list<PropertyList>, handler>::type &cgh, BinaryOperation combiner,
               PropertyList properties)
```

Reply:
I guess it's just general concern that adding another template parameter might cause some code to break.

@Pennycook, @rolandschulz: Didn't we see some previous example where adding a new template parameter to `accessor` caused something to break, even though the template parameter had a default value?  Maybe I'm misremembering?

Reply:
> That isn't the question. What I'm saying is that I'm not aware of any place in the codebase where we use such pattern instead of extra template parameter. Why should we do anything differently here?

FWIW, there are quite a few places (e.g., in the group algorithms) where we use `enable_if` on the return type instead of using an extra template parameter:

https://github.com/intel/llvm/blob/e9cbb8756e5cf578c912655933fc216d2af2ede0/sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp#L46-L49

> Didn't we see some previous example where adding a new template parameter to accessor caused something to break, even though the template parameter had a default value? Maybe I'm misremembering?

I think you're referring to this issue: https://github.com/intel/llvm/issues/10057.  Even if the same problem can't actually be triggered for this overload, the fact we hit this unexpected issue with `accessor` does make me think that it would be safer to avoid adding extra templates just in case.

Reply:
I agree with @aelovikov-intel, this is not an uncommon pattern for SFINAE on public APIs. If we believe this to be problematic, we may have to address this in other places as well, but the solution of SFINAE'ing an argument assumes that the methods have arguments, which holds in this case, but may not hold for all.

Reply:
I might be wrong about this, but if we wanted to be consistent, couldn't we always implement SFINAE using the return type?  Every function has a return type (but sometimes that type is `void`).

I appreciate it's a bit ugly to employ in a case like this where the implementation was previously relying on a deduced return type, but I think it's doable.  Is there a reason that we couldn't implement this as something like:

```c++
template <typename BufferT, typename BinaryOperation, typename PropertyList>
std::enable_if_t<detail::is_valid_reduction_prop_list<PropertyList>,
                 detail::reduction_t<BufferT, BinaryOperation, PropertyList>>
reduction(BufferT vars, handler &cgh, BinaryOperation combiner, PropertyList properties) { ... }
```

As an added bonus, all the SFINAE would be in a place where we could legally replace it with a `requires` statement when we switch to C++20:

```c++
template <typename BufferT, typename BinaryOperation, typename PropertyList>
requires ValidReductionPropertyList<PropertyList>
auto reduction(BufferT vars, handler &cgh, BinaryOperation combiner, PropertyList properties) { ... }
```

Reply:
I personally prefer SFINAE on return type, with the exceptions like this one when it's deduced :) ...and ctors.